diff --git a/cmake/pcl_find_boost.cmake b/cmake/pcl_find_boost.cmake
index cb26cce2a..a460ebc37 100644
--- a/cmake/pcl_find_boost.cmake
+++ b/cmake/pcl_find_boost.cmake
@@ -15,12 +15,12 @@ else()
 
 if(CMAKE_CXX_STANDARD MATCHES "14")
   # Optional boost modules
-  set(BOOST_OPTIONAL_MODULES serialization mpi system)
+  set(BOOST_OPTIONAL_MODULES serialization mpi)
   # Required boost modules
   set(BOOST_REQUIRED_MODULES filesystem iostreams)
 else()
   # Optional boost modules
-  set(BOOST_OPTIONAL_MODULES filesystem serialization mpi system)
+  set(BOOST_OPTIONAL_MODULES filesystem serialization mpi)
   # Required boost modules
   set(BOOST_REQUIRED_MODULES iostreams)
 endif()
diff --git a/cmake/pcl_pclconfig.cmake b/cmake/pcl_pclconfig.cmake
index 2d9b5e9e6..d8f5048c9 100644
--- a/cmake/pcl_pclconfig.cmake
+++ b/cmake/pcl_pclconfig.cmake
@@ -85,7 +85,7 @@ endforeach()
 
 #Boost modules
-set(PCLCONFIG_AVAILABLE_BOOST_MODULES "system iostreams")
+set(PCLCONFIG_AVAILABLE_BOOST_MODULES "iostreams")
 if(Boost_FILESYSTEM_FOUND)
   string(APPEND PCLCONFIG_AVAILABLE_BOOST_MODULES " filesystem")
 endif()
 if(Boost_SERIALIZATION_FOUND)
diff --git a/outofcore/CMakeLists.txt b/outofcore/CMakeLists.txt
index a27eedcdb..c15f8d92a 100644
--- a/outofcore/CMakeLists.txt
+++ b/outofcore/CMakeLists.txt
@@ -76,10 +76,10 @@ set(LIB_NAME "pcl_${SUBSYS_NAME}")
 
 if(BUILD_visualization)
   PCL_ADD_LIBRARY(${LIB_NAME} COMPONENT ${SUBSYS_NAME} SOURCES ${srcs} ${incs} ${impl_incs} ${visualization_incs})
-  target_link_libraries("${LIB_NAME}" pcl_common pcl_visualization Boost::system Boost::filesystem)
+  target_link_libraries("${LIB_NAME}" pcl_common pcl_visualization Boost::filesystem)
 else()
   PCL_ADD_LIBRARY(${LIB_NAME} COMPONENT ${SUBSYS_NAME} SOURCES ${srcs} ${incs} ${impl_incs})
-  target_link_libraries("${LIB_NAME}" pcl_common Boost::system Boost::filesystem)
+  target_link_libraries("${LIB_NAME}" pcl_common Boost::filesystem)
 endif()
 if(HAVE_CJSON)
   target_link_libraries("${LIB_NAME}" ${CJSON_LIBRARIES})
